### PR TITLE
Add flag to enable generation non-agg features

### LINKED ISSUE #267: Aggregated features and regular features cannot be in the same anchor

### PR DESCRIPTION
**Pull request by windoze**
## Description

Resolves #267 by adding a flag to enable materialization of non-agg features.
Normally materializing non-agg features doesn't output meaningful result, but the user can enable this by set the `allow_materialize_non_agg_feature` flag to `True`, the long name is on purpose to avoid misuse of this function.

## How was this PR tested?
Affected unit tests are updated.

## Does this PR introduce any user-facing changes?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure to clarify your proposed changes.

The materialization process is a grouping-by-key operation which mostly makes no sense to do over non-agg features as they don't have key for grouping.
In case the user really wants to do this, a flag is introduced to enable it.

### CODE REVIEW

File: feathr_project/feathr/client.py [hunk 1]
```diff
@@ -612,17 +613,31 @@ def _valid_materialize_keys(self, features: List[str], allow_empty_key=False):
...
         return True
     
-    def materialize_features(self, settings: MaterializationSettings, execution_configurations: Union[SparkExecutionConfiguration ,Dict[str,str]] = {}, verbose: bool = False):
+    def materialize_features(self, settings: MaterializationSettings, execution_configurations: Union[SparkExecutionConfiguration ,Dict[str,str]] = {}, verbose: bool = False, allow_materialize_non_agg_feature: bool = False):
         """Materialize feature data
 
         Args:
             settings: Feature materialization settings
             execution_configurations: a dict that will be passed to spark job when the job starts up, i.e. the "spark configurations". Note that not all of the configuration will be honored since some of the configurations are managed by the Spark platform, such as Databricks or Azure Synapse. Refer to the [spark documentation](https://spark.apache.org/docs/latest/configuration.html) for a complete list of spark configurations.
+            allow_materialize_non_agg_feature: Normally materializing non-aggregated features doesn't output meaningful results so it's forbidden, but if you really want to do this, set this to True.
```
Comment (xiaoyongzhu) [inline feathr_project/feathr/client.py:622]: slightly changing the description:

allow_materialize_non_agg_feature: Materializing non-aggregated features (the features without WindowAggTransformation) doesn't output meaningful results so it's by default set to False, but if you really want to materialize non-aggregated features, set this to True.

File: feathr_project/feathr/client.py [hunk 2]
```diff
@@ -612,17 +613,31 @@ def _valid_materialize_keys(self, features: List[str], allow_empty_key=False):
...
             raise RuntimeError(f"Invalid materialization features: {feature_list}, since they have different keys. Currently Feathr only supports materializing features of the same keys.")
         
+        if not allow_materialize_non_agg_feature:
+            # Check if there are non-aggregation features in the list
+            for fn in feature_list:
+                # Check over anchor features
+                for anchor in self.anchor_list:
+                    for feature in anchor.features:
+                        if feature.name == fn and not isinstance(feature.transform, WindowAggTransformation):
+                            raise RuntimeError(f"Feature {fn} is not an aggregation feature. Currently Feathr only supports materializing aggregation features.")
```
Comment (xiaoyongzhu) [inline feathr_project/feathr/client.py:635]: error message improvement: 

raise RuntimeError(f"Feature {fn} is not an aggregation feature. Currently Feathr only supports materializing aggregation features. If you want to materialize {fn}, please set allow_materialize_non_agg_feature to True.")

File: feathr_project/feathr/client.py [hunk 3]
```diff
@@ -612,17 +613,31 @@ def _valid_materialize_keys(self, features: List[str], allow_empty_key=False):
...
+                # Check over derived features
+                for feature in self.derived_feature_list:
+                    if feature.name == fn and not isinstance(feature.transform, WindowAggTransformation):
+                        raise RuntimeError(f"Feature {fn} is not an aggregation feature. Currently Feathr only supports materializing aggregation features.")
```
Comment (xiaoyongzhu) [inline feathr_project/feathr/client.py:639]: error message improvement: 

raise RuntimeError(f"Feature {fn} is not an aggregation feature. Currently Feathr only supports materializing aggregation features. If you want to materialize {fn}, please set allow_materialize_non_agg_feature to True.")

File: feathr_project/test/test_azure_spark_maven_e2e.py [hunk 4]
```diff
@@ -20,8 +24,37 @@ def test_feathr_online_store_agg_features():
 
     # The `feathr_runtime_location` was commented out in this config file, so feathr should use
     # Maven package as the dependency and `noop.jar` as the main file
-    client: FeathrClient = basic_test_setup(os.path.join(test_workspace_dir, "feathr_config_maven.yaml"))
+    client: FeathrClient = basic_test_setup(os.path.join(test_workspace_dir, "feathr_config_xch.yaml"))
```
Comment (xiaoyongzhu) [inline feathr_project/test/test_azure_spark_maven_e2e.py:27]: feathr_config_xch seems to be wrong

File: feathr_project/test/test_azure_spark_maven_e2e.py [hunk 5]
```diff
@@ -51,4 +84,6 @@ def test_feathr_online_store_agg_features():
     assert res['239'][0] != None
     assert res['239'][1] != None
     assert res['265'][0] != None
-    assert res['265'][1] != None
\ No newline at end of file
+    assert res['265'][1] != None
+
+test_feathr_online_store_agg_features()
```
Comment (xiaoyongzhu) [inline feathr_project/test/test_azure_spark_maven_e2e.py:89]: remove this?

**Comment by xiaoyongzhu** [comment]
Thanks @windoze ! I like this PR which gives user some flexibility. Some comments on the error message changes.

File: feathr_project/feathr/client.py (hunk 1 above)
Comment (windoze) [reply feathr_project/feathr/client.py]: Wording updated

File: feathr_project/feathr/client.py (hunk 3 above)
Comment (windoze) [reply feathr_project/feathr/client.py]: Wording updated

File: feathr_project/test/test_azure_spark_maven_e2e.py (hunk 4 above)
Comment (windoze) [reply feathr_project/test/test_azure_spark_maven_e2e.py]: Typo fixed

File: feathr_project/test/test_azure_spark_maven_e2e.py (hunk 5 above)
Comment (windoze) [reply feathr_project/test/test_azure_spark_maven_e2e.py]: Removed

File: feathr_project/feathr/client.py (hunk 2 above)
Comment (windoze) [reply feathr_project/feathr/client.py]: Wording updated

**Review by jainr — APPROVED** [review]
Looks good to me after making changes that Xiaoyong suggested.

**Review by edwincheung — DISMISSED** [review]

**Comment by xiaoyongzhu** [comment]
LGTM. @windoze please help fix the conflicts

**Comment by windoze** [comment]
> LGTM. @windoze please help fix the conflicts
Conflicts fixed.

**Review by blrchen — APPROVED** [review]

**Pull request merged**